py/server/deephaven/table.py

```diff
@@ -62,6 +69,102 @@ def _j_py_script_session() -> _JPythonScriptSession:
+
+    If a parameter or the return of the function is not annotated, the default 'O' - object type, will be used.
+    """
+    sig = inspect.signature(fn)
+
+    parameter_types = []
+    for n, p in sig.parameters.items():
+        try:
+            np_dtype = np.dtype(p.annotation if p.annotation else "object")
+            parameter_types.append(np_dtype)
+        except TypeError:
+            parameter_types.append(np.dtype("object"))
+
+    try:
+        return_type = np.dtype(sig.return_annotation if sig.return_annotation else "object")
+    except TypeError:
+        return_type = np.dtype("object")
+
+    np_type_codes = [np.dtype(p).char for p in parameter_types]
+    np_type_codes = [c if c in _numpy_type_codes else "O" for c in np_type_codes]
+    return_type_code = np.dtype(return_type).char
+    return_type_code = return_type_code if return_type_code in _numpy_type_codes else "O"
+
+    np_type_codes.extend(["-", ">", return_type_code])
+    return "".join(np_type_codes)
+
+
+class DhVectorize:
+    """A decorator to vectorize a Python function used in Deephaven query formulas and invoked on a row basis.
+
+    The result callable will accept arrays of the declared parameter data types (vs. scalar) and iterate through them
+    together and call the wrapped function on each iteration and pack the result of each call into an array and return
+    the result array upon exhausting the array parameters.
+    """
+
+    def __init__(self, fn: Callable):
+        self.callable = fn
+        self.signature = _encode_signature(fn)
+        self.return_type = self.signature[-1]
+        self.dh_vectorized = True
+        self._call_count = 0
+        self._arguments = [None] * (len(self.signature) - len("->?"))
+        self._chunk_size: int
+        if _test_vectorization:
+            global _vectorized_count
+            _vectorized_count += 1
+
+    def __call__(self, *args):
+        def is_chunked(v) -> bool:
+            if not isinstance(v, jpy.JType):
+                return False
+            if v.jclass.toString().startswith("class ["):
+                return True
+            else:
+                return False
+
+        # upon being called the first time, set up the return array and constant arguments for reuse in subsequent calls
+        if not self._call_count:
+            if len(args) != len(self._arguments) + 1:
+                raise ValueError(
+                    f"The number of arguments doesn't match the function signature. {len(args) - 1}, {self.signature}")
+
+            self._chunk_size = args[0]
+            self._chunk_result = np.empty(self._chunk_size, self.return_type)
+
+            for i, arg in enumerate(args[1:]):
+                if not is_chunked(arg):
+                    self._arguments[i] = np.full(self._chunk_size, arg)
+
+        self._chunk_size = args[0]
+        if self._chunk_size > len(self._chunk_result):
+            self._chunk_result = np.empty(self._chunk_size, self.return_type)
```

Comment:
This section has some weirdness and problems:
1) The result chunk allocation can probably be simplified to one location in the function, instead of two.
2) The non-cunked arguments are not resizing upon chunk size changes.